pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -0,0 +1,390 @@
+		return
+	}
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	klog.Infof("Waiting for caches to sync for %s", controllerName)
+	if !cache.WaitForCacheSync(stopCh, c.traceflowListerSynced) {
+		klog.Errorf("Unable to sync caches for %s", controllerName)
+		return
+	}
+	klog.Infof("Caches are synced for %s", controllerName)
+
+	go c.ofClient.ReceivePacketIn(c, stopCh)
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *Controller) addTraceflow(obj interface{}) {
+	tf := obj.(*opsv1alpha1.Traceflow)
+	klog.Infof("Processing Traceflow %s ADD event", tf.Name)
+	c.enqueueTraceflow(tf)
+}
+
+func (c *Controller) updateTraceflow(oldObj, curObj interface{}) {
+	tf := curObj.(*opsv1alpha1.Traceflow)
+	klog.Infof("Processing Traceflow %s UPDATE event", tf.Name)
+	c.enqueueTraceflow(tf)
+}
+
+func (c *Controller) deleteTraceflow(old interface{}) {
+	tf := old.(*opsv1alpha1.Traceflow)
+	klog.Infof("Processing Traceflow %s DELETE event", tf.Name)
+	c.deallocateTag(tf)
+}
+
+// worker is a long-running function that will continually call the processTraceflowItem function
+// in order to read and process a message on the workqueue.
+func (c *Controller) worker() {
+	for c.processTraceflowItem() {
+	}
+}
+
+// processTraceflowItem processes an item in the "traceflow" work queue, by calling syncTraceflow
+// after casting the item to a string (Traceflow name). If syncTraceflow returns an error, this
+// function handles it by updating phase to "ERROR". If syncTraceflow is successful, the Traceflow
+// is removed from the queue until we get notified of a new change. This function returns false if
+// and only if the work queue was shutdown (no more items will be processed).
+func (c *Controller) processTraceflowItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// We call Done here so the workqueue knows we have finished processing this item. We also
+	// must remember to call Forget if we do not want this work item being re-queued. For
+	// example, we do not call Forget if a transient error occurs, instead the item is put back
+	// on the workqueue and attempted again after a back-off period.
+	defer c.queue.Done(obj)
+
+	// We expect strings (Node name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen: enqueueTraceflow only enqueues strings.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncTraceflow(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again.
+		c.queue.Forget(key)
+	} else {
+		// If error occurs we log error and forget this item.
```

Comment:
I am a little confused even with the comment here. 
Typically, we do not clear the rate limit by using Forget if there is any error. Could you explain more why we need to forget in this case?
If we do not need the rate limit here, then I guess we do not need RatelimitQueue here?

Reply:
I some previous versions, retry is used here, and we removed retry in current version. I think we can keep the RatelimitQueue since the overhead is acceptable and we may add retry in future.
I'll update comment of this function.

Reply:
Removed c.queue.Forget(key) for error case.